src/core/qgsmaplayerlegend.cpp

```diff
@@ -586,13 +590,76 @@ QList<QgsLayerTreeModelLegendNode *> QgsDefaultMeshLayerLegend::createLayerTreeM
         settings.colorRampShader().legendSymbologyItems( items );
         for ( const QPair< QString, QColor > &item : items )
         {
-          nodes << new QgsRasterSymbolLegendNode( nodeLayer, item.second, item.first );
+          nodes << new QgsRasterSymbolLegendNode( nodeLayer, item.second, item.first, nullptr, false,
+                                                  QStringLiteral( "scalarLegend" ) + QUuid::createUuid().toString(),
+                                                  scalarNameKey
+                                                );
         }
         break;
       }
     }
   }
 
+  if ( indexVector > -1 )
+  {
+    const QgsMeshRendererVectorSettings settings = rendererSettings.vectorSettings( indexVector );
+    const QString vectorNameKey = QStringLiteral( "vectorName" );
+    switch ( settings.coloringMethod() )
+    {
+      case QgsInterpolatedLineColor::ColoringMethod::SingleColor:
+      {
+        const QColor arrowColor = settings.color();
+        const QIcon vectorIcon = QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/meshvectors.svg" ), arrowColor, arrowColor );
+        nodes << new QgsSimpleLegendNode( nodeLayer, mLayer->datasetGroupMetadata( indexVector ).name(),
+                                          vectorIcon, nullptr, vectorNameKey );
+        break;
+      }
+      case QgsInterpolatedLineColor::ColoringMethod::ColorRamp:
+      {
+        const QIcon vectorIcon = QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/meshvectors.svg" ) );
+        nodes << new QgsSimpleLegendNode( nodeLayer, mLayer->datasetGroupMetadata( indexVector ).name(),
+                                          vectorIcon, nullptr, vectorNameKey );
+        const QgsColorRampShader shader = settings.colorRampShader();
+        switch ( shader.colorRampType() )
+        {
+          case QgsColorRampShader::Interpolated:
+            if ( !shader.legendSettings() || shader.legendSettings()->useContinuousLegend() )
+            {
+              // for interpolated shaders we use a ramp legend node
+              if ( !shader.colorRampItemList().isEmpty() )
+              {
+                nodes << new QgsColorRampLegendNode( nodeLayer, shader.createColorRamp(),
+                                                     shader.legendSettings() ? *shader.legendSettings() : QgsColorRampLegendNodeSettings(),
+                                                     shader.minimumValue(),
+                                                     shader.maximumValue(),
+                                                     nullptr,
+                                                     QStringLiteral( "vectorLegend" ),
+                                                     vectorNameKey );
+              }
+              break;
+            }
+            [[fallthrough]];
```

Comment:
Can you avoid the fallthrough? If something is to be run in all cases then I'd rather have it outside the switch statement.

Reply:
The fallthroughs are not mine. Multiple fallthroughs throughout the project were modified/added by @nyalldawson on PR #55018 .  If the fallthroughs are an issue, then there should be a concerted effort to remove them in a separate PR. But I am of the opinion that this does not fall under the purview of this PR.

Reply:
I was referring to the actual logic that allows to fall through from one switch case to the next, however I now saw that this is copied from the `scalar` code above, so better keep it as is indeed.